pkg/spoke/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,148 @@
+}
+
+func NewFinalizeController(
+	recorder events.Recorder,
+	spokeDynamicClient dynamic.Interface,
+	manifestWorkClient workv1client.ManifestWorkInterface,
+	manifestWorkInformer workinformer.ManifestWorkInformer,
+	manifestWorkLister worklister.ManifestWorkNamespaceLister,
+) factory.Controller {
+
+	controller := &FinalizeController{
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		spokeDynamicClient: spokeDynamicClient,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, manifestWorkInformer.Informer()).
+		WithSync(controller.sync).ToController("ManifestWorkFinalizer", recorder)
+}
+
+func (m *FinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	manifestWorkName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling ManifestWork %q", manifestWorkName)
+
+	manifestWork, err := m.manifestWorkLister.Get(manifestWorkName)
+	if errors.IsNotFound(err) {
+		// work  not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	return m.syncManifestWork(ctx, manifestWork)
+}
+
+// syncManifestWork ensures that when a manifestwork has been deleted, everything it created is also deleted before removing
+// the finalizer
+func (m *FinalizeController) syncManifestWork(ctx context.Context, originalManifestWork *workapiv1.ManifestWork) error {
+	manifestWork := originalManifestWork.DeepCopy()
+
+	// no work to do until we're deleted
+	if manifestWork.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// don't do work if the finalizer is not present
+	found := false
+	for i := range manifestWork.Finalizers {
+		if manifestWork.Finalizers[i] == manifestWorkFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	// Work is deleting, we remove its related resources on spoke cluster
+	if errs := m.cleanupResourceOfWork(manifestWork); len(errs) != 0 {
+		return utilerrors.NewAggregate(errs)
+	}
+
+	// We consider the case of deletion of created resources, with finalization still to come, as sufficient to remove the finalizer
+	// We do this because resources cannot be un-deleted.  This means that deletion is inevitable.
+	// Also, since we don't track UIDs, we have no reliable way of know when "this" particular resource has been removed as
+	// compared with a case where this controller deletes it and another controller (or manifestwork) creates it.
+
+	removeFinalizer(manifestWork, manifestWorkFinalizer)
+	_, err := m.manifestWorkClient.Update(ctx, manifestWork, metav1.UpdateOptions{})
+	return err
+}
+
+func (m *FinalizeController) cleanupResourceOfWork(work *workapiv1.ManifestWork) []error {
+	klog.V(4).Infof("cleaning up %q", work.Name)
+
+	errs := []error{}
+
+	// TODO this can later be based on a list of all resources created by this manifest work
```

Comment:
i think the intension is we should have another filed in manifestwork.Status to track all the resources this manifestwork have created, in addition to the resources that this manifestwork is currently managed?

Tracking resources create by the manifestwork also can be used in other scenarios, such as:
1. the name or namespace of a manifest in the manifestwork.Spec.Workload is changed.
2. a manifest is removed from manifestwork.Spec.Workload.

Reply:
> 2\. a manifest is removed from manifestwork.Spec.Workload.

Yes.  that's what I'm thinking.  An example here: https://github.com/open-cluster-management/api/pull/26

Reply:
+1.